graph/src/data/graphql/validation_hash.rs

```diff
@@ -0,0 +1,415 @@
+        // Operation -> Fragment
+        _ => definition_kind_ordering(a).cmp(&definition_kind_ordering(b)),
+    }
+}
+
+fn compare_selections<'a>(a: &q::Selection, b: &q::Selection) -> Ordering {
+    match (a, b) {
+        // Sort fields by name
+        (q::Selection::Field(a), q::Selection::Field(b)) => a.name.cmp(&b.name),
+        // Sort fragments by name
+        (q::Selection::FragmentSpread(a), q::Selection::FragmentSpread(b)) => {
+            a.fragment_name.cmp(&b.fragment_name)
+        }
+        _ => {
+            let a_ordering = selection_kind_ordering(a);
+            let b_ordering = selection_kind_ordering(b);
+            a_ordering.cmp(&b_ordering)
+        }
+    }
+}
+
+fn compare_arguments<'a>(a: &(String, q::Value), b: &(String, q::Value)) -> Ordering {
+    a.0.cmp(&b.0)
+}
+
+fn compare_variable_definitions<'a>(
+    a: &q::VariableDefinition,
+    b: &q::VariableDefinition,
+) -> Ordering {
+    a.name.cmp(&b.name)
+}
+
+/// Assigns an order to different variants of Selection
+fn selection_kind_ordering<'a>(selection: &q::Selection) -> u8 {
+    match selection {
+        q::Selection::FragmentSpread(_) => 1,
+        q::Selection::InlineFragment(_) => 2,
+        q::Selection::Field(_) => 3,
+    }
+}
+
+/// Assigns an order to different variants of Definition
+fn definition_kind_ordering<'a, T: q::Text<'a>>(definition: &q::Definition<'a, T>) -> u8 {
+    match definition {
+        q::Definition::Operation(_) => 1,
+        q::Definition::Fragment(_) => 2,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use graphql_parser::parse_query;
+
+    #[test]
+    fn sorted() {
+        const Q1: &str = "{ things(first: 10, limit: 20) { e { ... on E { e1 e3 e2 } } f { ... F } a c b d } } fragment F on F { f1 f3 f2 }";
+        const Q2: &str = "{ things(limit: 20, first: 10) { a c b d f { ...F } e { ... on E { e3 e1 e2 } } } } fragment F on F { f2 f3 f1 }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+
+        assert_eq!(validation_hash(&q1), validation_hash(&q2));
+    }
+
+    #[test]
+    fn do_not_sort_inline_fragments() {
+        const Q1: &str = "      { things { ... on ThingA { a }    ... on ThingB { b }    } }";
+        const Q2: &str = "      { things { ... on ThingB { b }    ... on ThingA { a }    } }";
+        const Q3: &str = "query { things { ... on ThingA { c: a } ... on ThingB { c: b } } }";
+        const Q4: &str = "query { things { ... on ThingB { c: b } ... on ThingA { c: a } } }";
+        let q1 = parse_query(Q1)
+            .expect("q1 is syntactically valid")
+            .into_static();
+        let q2 = parse_query(Q2)
+            .expect("q2 is syntactically valid")
+            .into_static();
+        let q3 = parse_query(Q3)
```

Comment:
Nit: By making a helper function with `#[track_caller]` that does both `parse_query` and `validation_hash` you can make all the test cases more succinct.